lib/maths/CCalendarCyclicTest.cc

```diff
@@ -0,0 +1,328 @@
+const std::string ERRORS_6_4_TAG("e");
+// Version < 6.4
+const std::string ERROR_QUANTILES_OLD_TAG("a");
+// Everything else gets default initialised.
+
+const std::string DELIMITER{","};
+const core_t::TTime SIZE{155};
+const core_t::TTime BUCKET{core::constants::DAY};
+const core_t::TTime WINDOW{SIZE * BUCKET};
+const core_t::TTime TIME_ZONE_OFFSETS[]{0};
+
+//! The percentile of a large error.
+const double LARGE_ERROR_PERCENTILE{98.5};
+//! The minimum number of repeats to test a feature.
+const unsigned int MINIMUM_REPEATS{4};
+//! The maximum significance to accept a feature.
+const double MAXIMUM_SIGNIFICANCE{0.01};
+}
+
+CCalendarCyclicTest::CCalendarCyclicTest(double decayRate)
+    : m_DecayRate{decayRate}, m_ErrorQuantiles{CQuantileSketch::E_Linear, 20},
+      m_CurrentBucketTime{0}, m_CurrentBucketIndex{0} {
+    static const SSetTimeZone timezone("GMT");
+    TErrorStatsVec stats(SIZE);
+    this->deflate(stats);
+}
+
+bool CCalendarCyclicTest::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) {
+    TErrorStatsVec errors;
+    if (traverser.name() == VERSION_6_4_TAG) {
+        while (traverser.next()) {
+            const std::string& name = traverser.name();
+            RESTORE(ERROR_QUANTILES_6_4_TAG,
+                    traverser.traverseSubLevel(boost::bind(&CQuantileSketch::acceptRestoreTraverser,
+                                                           &m_ErrorQuantiles, _1)))
+            RESTORE_BUILT_IN(CURRENT_BUCKET_TIME_6_4_TAG, m_CurrentBucketTime)
+            RESTORE_BUILT_IN(CURRENT_BUCKET_INDEX_6_4_TAG, m_CurrentBucketIndex)
+            RESTORE(CURRENT_BUCKET_ERROR_STATS_6_4_TAG,
+                    m_CurrentBucketErrorStats.fromDelimited(traverser.value()))
+            RESTORE(ERRORS_6_4_TAG,
+                    core::CPersistUtils::restore(ERRORS_6_4_TAG, errors, traverser))
+        }
+    } else {
+        do {
+            const std::string& name = traverser.name();
+            RESTORE(ERROR_QUANTILES_OLD_TAG,
+                    traverser.traverseSubLevel(boost::bind(&CQuantileSketch::acceptRestoreTraverser,
+                                                           &m_ErrorQuantiles, _1)))
+        } while (traverser.next());
+        errors.resize(SIZE);
+    }
+    this->deflate(errors);
+    return true;
+}
+
+void CCalendarCyclicTest::acceptPersistInserter(core::CStatePersistInserter& inserter) const {
+    inserter.insertValue(VERSION_6_4_TAG, "");
+    inserter.insertLevel(ERROR_QUANTILES_6_4_TAG,
+                         boost::bind(&CQuantileSketch::acceptPersistInserter,
+                                     &m_ErrorQuantiles, _1));
+    inserter.insertValue(CURRENT_BUCKET_TIME_6_4_TAG, m_CurrentBucketTime);
+    inserter.insertValue(CURRENT_BUCKET_INDEX_6_4_TAG, m_CurrentBucketIndex);
+    inserter.insertValue(CURRENT_BUCKET_ERROR_STATS_6_4_TAG,
+                         m_CurrentBucketErrorStats.toDelimited());
+    TErrorStatsVec errors{this->inflate()};
```

Comment:
so the compressed buffer is only an internal detail and will not be persisted/restored?

would be good to document that somewhere